gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -83,6 +95,211 @@
     }
   }
 
+  /**
+   * Group workUnits into multiWorkUnits. Each workUnit corresponds to a (topic, partition).
+   * The goal is to group the workUnits as evenly as possible into numOfMultiWorkunits multiWorkunits,
+   * while preferring to put partitions of the same topic into the same multiWorkUnits (in order to
+   * avoid generating many small mapper output files).
+   *
+   * The algorithm is to first group workUnits into approximately 3 * numOfMultiWorkunits groups using
+   * best-fit-decreasing, such that partitions of the same topic are in the same group. Then, 
+   * assemble these groups into multiWorkunits using worst-fit-decreasing.
+   *
+   * The reason behind 3 is that the worst-fit-decreasing algorithm should work well if the number of
+   * items is more than 3 times the number of bins.
+   *
+   * The input workUnits are partitioned by topic.
+   */
+  private List<WorkUnit> getMultiWorkunits(List<List<WorkUnit>> workUnits, int numOfMultiWorkunits) {
+    Preconditions.checkArgument(numOfMultiWorkunits >= 1);
+
+    List<List<WorkUnitWithEstSize>> workUnitsWithEstSize = getWorkUnitsWithEstSize(workUnits);
```

Comment:
Instead of creating this new class `WorkUnitWithEstSize`, why not simply setting a property in the `WorkUnit`?